### PR TITLE
Create approve templates

### PR DESCRIPTION
**Pull request by pauloancheta**
depends on #41 

## Usage
- Change `ApprovedTemplateLocation` to the full s3 url with key of the template to be approved
- Run command: `iidy approve-template stack-args.yaml`

### Notes:
I could also change it to `iidy approve-template https://s3.amazon.com...yaml.pending` but to add `--profile` functionality is more tedious.
Ideally, it would be using `iidy approve-template hash.yaml.pending` but the command would not know which profile to use and which bucket to look at.
 
Added a flag `--approve` to approve the template. Running the `iidy approve-template ...` command should diff the changes first. I have opted not to use a `readline` to precent fat fingering answers.

## TODO:
- [x] Switch base to master once #41 has been approved and merged

### CODE REVIEW

File: src/approval/index.ts [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+        diff.forEach(function(part){
+          if (part.added) {
+            colorizedString = colorizedString + cli.green(part.value)
+          } else if (part.removed) {
+            colorizedString = colorizedString + cli.red(part.value)
+          }
+        });
+        console.log(colorizedString);
+        console.log("Do these changes look good for you? Add a flag `--approve true` to approve");
```
Comment (jpb) [inline src/approval/index.ts]: Could this be interactive? Prompt for y/n similar to [this](https://github.com/unbounce/iidy/blob/8a798337fd3cf757c47e832d295f92eefc9a60ba/src/cfn/index.ts#L1315).

File: src/approval/index.ts [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+        return 0
+      }
+
+      // create a new pending file
```
Comment (jpb) [inline src/approval/index.ts]: This should probably be in the `else` branch of the conditional above (even though there is a `return`).

File: src/approval/index.ts [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+      await s3.putObject({
+        Body: pendingTemplate,
+        Bucket: s3Url.bucket,
+        Key: s3Url.key.replace(/\.pending$/, "")
+      }).promise();
+      logSuccess("Created a new cfn-template.yaml.");
```
Comment (jpb) [inline src/approval/index.ts]: I think you should log this and "Updated latest" at debug level.

File: src/approval/index.ts [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    }
+  }
+
+  return 0;
+}
+
+export async function approveTemplate(argv: Arguments): Promise<number> {
+  await configureAWS(argv.profile, "us-east-1");
+  const s3 = new S3();
+  const s3Url = parseS3HttpUrl(argv.filename);
```
Comment (jpb) [inline src/approval/index.ts]: The same as https://github.com/unbounce/iidy/pull/41#discussion_r160296403, this is a misuse of `parseS3HttpUrl`

File: src/approval/index.ts [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+      // command has to be run again with the flag to approve changes
+      if (argv.approve === false) {
+        const previouslyApprovedTemplate = await s3.getObject({
+          Bucket: s3Url.bucket,
+          Key: `${bucketDir}/latest.yaml`
+        }).promise()
+          .then((template) => template.Body)
+          .catch((e) => {
+            if (e.code !== "NoSuchKey") {
+              return Promise.reject(e);
```
Comment (jpb) [inline src/approval/index.ts]: 👌

File: src/cfn/index.ts [hunk 6]
```diff
@@ -1515,6 +1516,7 @@ export async function convertStackToIIDY(argv: Arguments): Promise<number> {
   const stackArgs: StackArgs = {
     Template: './cfn-template.yaml',
     StackName: StackNameArg,
+    ApprovedTemplateLocation: '',
```
Comment (tavisrudd) [inline src/cfn/index.ts]: Shouldn't the default here be `undefined`?

File: src/approval/index.ts [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+    const templatePath = path.resolve(path.dirname(argv.argsfile), stackArgs.Template);
+    const cfnTemplate = await fs.readFileSync(templatePath);
+
+    const s3Url = url.parse(stackArgs.ApprovedTemplateLocation);
+    const s3Path = s3Url.path ? s3Url.path : "";
+    const s3Bucket = s3Url.hostname ? s3Url.hostname : "";
+
+    const fileName = new Md5().appendStr(cfnTemplate.toString()).end().toString()
+    const fullFileName = `${fileName}${path.extname(stackArgs.Template)}.pending`
```
Comment (tavisrudd) [inline src/approval/index.ts]: semicolons on this and line 30.

File: src/approval/index.ts [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+            return Promise.reject(e);
+          }
+          return Buffer.from("");
+        });
+
+      const diff = jsdiff.diffLines(
+        previouslyApprovedTemplate!.toString(),
+        pendingTemplate!.toString(),
+      );
+      let colorizedString = ""
```
Comment (tavisrudd) [inline src/approval/index.ts]: semicolon

Comment (tavisrudd) [reply src/approval/index.ts]: same in other spots in this file

**Comment by tavisrudd** [comment]
Let's squash the commits prior to merge and then rebase #45.

**Pull request merged**